src/engine/GroupByHashMapOptimization.h

```diff
@@ -0,0 +1,137 @@
+    }
+    auto impl = [&ctx]<typename X, typename Y>(const X& x, const Y& y) {
+      return toBoolNotUndef(
+          sparqlExpression::compareIdsOrStrings<
+              Comp, valueIdComparators::ComparisonForIncompatibleTypes::
+                        CompareByType>(x, y, ctx));
+    };
+    if (std::visit(impl, value, currentValue_)) currentValue_ = value;
+  }
+  [[nodiscard]] ValueId calculateResult(LocalVocab* localVocab) const {
+    auto valueIdResultGetter = [](ValueId id) { return id; };
+    auto stringResultGetter = [localVocab](const std::string& str) {
+      auto localVocabIndex = localVocab->getIndexAndAddIfNotContained(str);
+      return ValueId::makeFromLocalVocabIndex(localVocabIndex);
+    };
+    return std::visit(ad_utility::OverloadCallOperator(valueIdResultGetter,
+                                                       stringResultGetter),
+                      currentValue_);
+  }
+};
+
+using MinAggregationData =
+    ExtremumAggregationData<valueIdComparators::Comparison::LT>;
+using MaxAggregationData =
+    ExtremumAggregationData<valueIdComparators::Comparison::GT>;
+
+// Data to perform the SUM aggregation using the HashMap optimization.
+struct SumAggregationData {
+  using ValueGetter = sparqlExpression::detail::NumericValueGetter;
+  bool error_ = false;
+  double sum_ = 0;
+  void addValue(auto&& value, const sparqlExpression::EvaluationContext* ctx) {
+    auto val = ValueGetter{}(AD_FWD(value), ctx);
+    auto intValueAdder = [this](int64_t intValue) {
+      sum_ += static_cast<double>(intValue);
+    };
+    auto doubleValueAdder = [this](double doubleValue) { sum_ += doubleValue; };
+    auto notNumericAdder = [this](sparqlExpression::detail::NotNumeric) {
+      error_ = true;
+    };
+    std::visit(ad_utility::OverloadCallOperator(intValueAdder, doubleValueAdder,
+                                                notNumericAdder),
+               val);
+  };
+  [[nodiscard]] ValueId calculateResult(
+      [[maybe_unused]] const LocalVocab* localVocab) const {
+    if (error_)
+      return ValueId::makeUndefined();
+    else
+      return ValueId::makeFromDouble(sum_);
+  }
+};
+
+// Data to perform GROUP_CONCAT aggregation using the HashMap optimization.
+struct GroupConcatAggregationData {
+  using ValueGetter = sparqlExpression::detail::StringValueGetter;
+  std::string currentValue_;
+  std::string separator_;
+  void addValue(auto&& value, const sparqlExpression::EvaluationContext* ctx) {
+    auto val = ValueGetter{}(AD_FWD(value), ctx);
+    if (val.has_value()) {
+      if (!currentValue_.empty()) currentValue_.append(separator_);
+      currentValue_.append(val.value());
+    }
+  }
+  [[nodiscard]] ValueId calculateResult(LocalVocab* localVocab) const {
+    auto localVocabIndex =
+        localVocab->getIndexAndAddIfNotContained(currentValue_);
+    return ValueId::makeFromLocalVocabIndex(localVocabIndex);
+  }
+
+  explicit GroupConcatAggregationData(std::string separator)
```

Comment:
In this file, you can add the `// _______________________________` coments
before the functions and constructors to increase the readability.